language/move-core/types/src/language_storage.rs

```diff
@@ -79,6 +103,153 @@ impl<'a> From<&'a TypeTag> for SafeTypeTag<'a> {
     }
 }
 
+// u8 -> depth or nesting level
+struct TypeTagVisitor(u8);
+
+use serde::de::{
+    self, DeserializeSeed, Deserializer, EnumAccess, MapAccess, SeqAccess, VariantAccess, Visitor,
+};
+
+impl<'de> Visitor<'de> for TypeTagVisitor {
+    type Value = TypeTag;
+
+    fn expecting(&self, formatter: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        formatter.write_fmt(format_args!("`{}`", TYPE_TAG_VARIANTS[0]))?;
+        for value in &TYPE_TAG_VARIANTS[1..] {
+            formatter.write_fmt(format_args!(", `{}`", value))?;
+        }
+        Ok(())
+    }
+
+    fn visit_enum<A>(self, data: A) -> Result<Self::Value, A::Error>
+    where
+        A: EnumAccess<'de>,
+    {
+        let (variant_type, variant) = data.variant::<DeTypeTag>()?;
+        match variant_type {
+            DeTypeTag::Bool => {
+                variant.unit_variant()?;
+                Ok(TypeTag::Bool)
+            }
+            DeTypeTag::U8 => {
+                variant.unit_variant()?;
+                Ok(TypeTag::U8)
+            }
+            DeTypeTag::U64 => {
+                variant.unit_variant()?;
+                Ok(TypeTag::U64)
+            }
+            DeTypeTag::U128 => {
+                variant.unit_variant()?;
+                Ok(TypeTag::U128)
+            }
+            DeTypeTag::Address => {
+                variant.unit_variant()?;
+                Ok(TypeTag::Address)
+            }
+            DeTypeTag::Signer => {
+                variant.unit_variant()?;
+                Ok(TypeTag::Signer)
+            }
+            DeTypeTag::Vector => Ok(TypeTag::Vector(Box::new(
+                variant.newtype_variant_seed(NestedTypeTag(self.0))?,
```

Comment:
Why not just make `self.0` be `self.nest` or something.

Reply:
I hate extra code :)